### PR TITLE
feat(forge): add contract ignore list for gas reports

### PR DESCRIPTION
**Pull request by derch28**
## Motivation

See #2464 

## Solution

Create a new Config field called `gas_reports_ignore` and a new field in `GasReport` called `ignore`. (Default value for `gas_reports_ignore` is an empty Vec&lt;String&gt;.)

Pass `gas_reports_ignore` as an argument to function `GasReport::new` so `GasReport`s are built with an ignore list.

The initialization of the variable `report_contract` decides if the contract has to be reported, using this logic:
image

Were:
- a: self.report_for.contains(&contract_name)
- b: self.ignore.contains(&contract_name)
- c: self.report_for.is_empty()
- d: self.report_for.contains(&"*".to_string())

In simpler terms: Report the contract every time it's listed on `report_for`. Report all contracts if `report_for` is either empty or contains "*". If you're reporting all contracts but one of them is listed on `ignore`, don't report that one.

If the user listed the contract in 'gas_reports' (the foundry.toml field) a report for the contract is generated even if it's listed in the ignore list. This is addressed this way because getting a report you don't expect is preferable than not getting one you expect. A warning is printed to stderr indicating the "double listing".

### CODE REVIEW

File: forge/src/gas_report.rs [hunk 1]
```diff
@@ -30,12 +31,14 @@ pub struct GasInfo {
...
-    pub fn new(report_for: Vec<String>) -> Self {
-        Self { report_for, ..Default::default() }
+    pub fn new(report_for: Vec<String>, ignore: Vec<String>) -> Self {
+        Self { report_for, ignore, ..Default::default() }
     }
 
     pub fn analyze(&mut self, traces: &[(TraceKind, CallTraceArena)]) {
-        let report_for_all = self.report_for.is_empty() || self.report_for.iter().any(|s| s == "*");
+        let report_for_all = (self.report_for.is_empty() ||
+            self.report_for.iter().any(|s| s == "*")) &
```
Comment (onbjerg) [inline forge/src/gas_report.rs:40]:
```suggestion
            self.report_for.iter().any(|s| s == "*")) &&
```

File: forge/src/gas_report.rs [hunk 2]
```diff
@@ -58,7 +61,11 @@ impl GasReport {
             let report_for = self
                 .report_for
                 .iter()
-                .any(|s| s == name.rsplit(':').next().unwrap_or(name.as_str()));
+                .any(|s| s == name.rsplit(':').next().unwrap_or(name.as_str())) &
```
Comment (onbjerg) [inline forge/src/gas_report.rs:64]:
```suggestion
                .any(|s| s == name.rsplit(':').next().unwrap_or(name.as_str())) &&
```

Comment (onbjerg) [inline forge/src/gas_report.rs:64]: We can replace the double iterators with something like:

```rs
let contract_name = name.rsplit(':').next().unwrap_or(name.as_str());
let report_for = self.report_for.contains(contract_name) && !self.ignore.contains(contract_name);
```

**Review by onbjerg — COMMENTED** [review]
Some minor nits, otherwise good to me so far :)

File: forge/src/gas_report.rs (hunk 2 above)
Comment (derch28) [reply forge/src/gas_report.rs]: Whoops, you're right. Just committed what you suggested, with a little change to satisfy the argument type of `.contains()`.

**Comment by onbjerg** [comment]
Let me know if something is blocking you :)

**Comment by derch28** [comment]
As I mentioned [here](https://github.com/foundry-rs/foundry/issues/2464#issuecomment-1200055646), I need help testing it because I don't know how to write tests for it. So I'd really appreciate if you let me know how can I learn to test new features :)

**Comment by onbjerg** [comment]
@derch28 I think the best way currently would be to add a new `forgetest!` in `cli/tests/it/cmd.rs` and checking that the output of the command contains or does not contain what you expect - be careful not to match on the entirey of the output (use `String::contains` instead) since some of the output may change between each run.

A good starting point is reading some of the tests in there, figuring out what is closest to what you want to do (e.g. setting `foundry.toml`, adding a simple contract, running `forge test --gas-report`)

**Comment by derch28** [comment]
@onbjerg Followed what you said and managed to write some tests :)

While testing, I realized that there was a simpler way to solve the issue. (And that even allowed me to delete a function that was no longer needed.)

So, given that all the previous commits don't represent any logical change (because the PR-ready code was all made in the last one) I'm squashing them and force-pushing the result.

Should I edit the "Solution" Section of  [my first comment](https://github.com/foundry-rs/foundry/pull/2528#issue-1323359692) and explain how the new solution works there?

File: forge/src/gas_report.rs [hunk 3]
```diff
@@ -54,12 +50,20 @@ impl GasReport {
...
-                .any(|s| s == name.rsplit(':').next().unwrap_or(name.as_str()));
-            if report_for || report_for_all {
+            let contract_name = name.rsplit(':').next().unwrap_or(name.as_str()).to_string();
+            // If the user listed the contract in 'gas_reports' (the foundry.toml field) a
+            // report for the contract is generated even if it's listed in the ignore
+            // list. This is addressed this way because getting a report you don't expect is
+            // preferable than not getting one you expect. A warning is printed to stderr
+            // indicating the "double listing".
+            if self.report_for.contains(&contract_name) && self.ignore.contains(&contract_name) {
+                eprintln!("\x1b[93mWarning: \x1b[0m {} is listed in both 'gas_reports' and 'gas_reports_ignore'.", contract_name);
```
Comment (onbjerg) [inline forge/src/gas_report.rs:60]: use `yansi` here since we can disable color output for users who do not support color (e.g. windows users): https://docs.rs/yansi/latest/yansi/

File: forge/src/gas_report.rs [hunk 4]
```diff
@@ -54,12 +50,20 @@ impl GasReport {
...
+            }
+            let report_contract = (!self.ignore.contains(&contract_name) &&
+                self.report_for.iter().any(|s| s == "*")) ||
```
Comment (onbjerg) [inline forge/src/gas_report.rs:63]:
```suggestion
                self.report_for.contains("*") ||
```

**Review by onbjerg — APPROVED** [review]
looks good :) two minor nits, nice one on the clean up

File: cli/tests/it/cmd.rs [hunk 5]
```diff
@@ -861,3 +861,400 @@ contract MyTokenCopy is MyToken {
...
+            "ContractTwo".to_string(),
+            "ContractThree".to_string(),
+        ]),
+        gas_reports_ignore: (vec!["ContractThree".to_string()]),
+        ..Default::default()
+    });
+    cmd.forge_fuse();
+    let third_out = cmd.arg("test").arg("--gas-report").stdout();
+    assert!(third_out.contains("foo") && third_out.contains("bar") && third_out.contains("baz"));
+    // cmd.arg("test").arg("--gas-report").print_output();
```
Comment (mattsse) [inline cli/tests/it/cmd.rs:1259]:
```suggestion
```

**Review by mattsse — APPROVED** [review]
thx for this,

lgtm

File: cli/tests/it/cmd.rs [hunk 6]
```diff
@@ -861,3 +861,399 @@ contract MyTokenCopy is MyToken {
...
+            "ContractOne".to_string(),
+            "ContractTwo".to_string(),
+            "ContractThree".to_string(),
+        ]),
+        gas_reports_ignore: (vec!["ContractThree".to_string()]),
+        ..Default::default()
+    });
+    cmd.forge_fuse();
+    let third_out = cmd.arg("test").arg("--gas-report").stdout();
+    assert!(third_out.contains("foo") && third_out.contains("bar") && third_out.contains("baz"));
```
Comment (gakonst) [inline cli/tests/it/cmd.rs:1258]: `ContractThree::baz` doesn't look ignored here?

Comment (derch28) [reply cli/tests/it/cmd.rs:1258]: In this case, "ContractThree" was added to the report list and the ignore list in the same run.

I explained how I decided to handle this behavior in forge/src/gas_report.rs line 53:

"If the user listed the contract in 'gas_reports' (the foundry.toml field) a report for the contract is generated even if it's listed in the ignore list. This is addressed this way because getting a report you don't expect is preferable than not getting one you expect. A warning is printed to stderr indicating the "double listing"."
